server/src-lib/Data/Parser/JSONPath.hs

```diff
@@ -0,0 +1,51 @@
+module Data.Parser.JSONPath
+  ( parseJSONPath
+  , JSONPathElement(..)
+  ) where
+
+import           Control.Applicative  ((<|>))
+import           Data.Attoparsec.Text
+import           Prelude
+
+import qualified Data.Text            as T
+
+data JSONPathElement
+  = Key T.Text
+  | Index !Int
+  deriving (Show, Eq)
+
+firstPathElementChar :: Parser Char
+firstPathElementChar = letter
+  <|> satisfy (`elem` ("_$" :: String))
+  <?> "The first character must be a letter, an underscore (_) or a dollar sign ($)."
+
+pathElement :: Parser JSONPathElement
+pathElement = do
+  firstChar <- firstPathElementChar
+  remain <- many' ( letter
+                <|> digit
+                <|> satisfy (`elem` ("-_$" :: String))
+                )
+  return $ Key $ T.pack (firstChar:remain)
+
+childPathElement :: Parser JSONPathElement
+childPathElement = skip (== '.') *> pathElement
+
+indexPathElement :: Parser JSONPathElement
+indexPathElement = do
+  skip (== '[')
+  content <- many1' digit <?> "Invalid array JSON path. Expected digits, got empty or valid characters"
+  skip (== ']')
+  return $ Index (read content)
```

Comment:
`read` is error prone. Use `readMaybe` instead.

Reply:
Because the result can be multiple digit, the index will be larger than 10. `many1' digit ~ Parser String`, not Char

Reply:
Oh! I see. But I'm assuming we're missing few corner cases here. If content is of type `[Char]` , what happens if it is empty array? This occurs when we trying to parse `name.addresses[]` (any `[]`).

Reply:
`many1` will check empty for it, from the documentation. However, the error is not friendly, so I need to improve a bit

Reply:
We don't encourage using pure code functions prone to errors. `read` functions throws `error` exception, though `digit` parser guarantees a `number`. I suggest you use `readMaybe` and throw parse error when it is `Nothing`